keps/sig-autoscaling/20181106-in-place-update-of-pod-resources.md

```diff
@@ -303,10 +303,14 @@ Pod v1 core API:
 Admission Controllers: LimitRanger, ResourceQuota need to support Pod Updates:
 * for ResourceQuota it should be enough to change podEvaluator.Handler
   implementation to allow Pod updates,
-* Pod's Spec.Containers[i].ResourcesAllocated should be used to be in line with
-  current ResourceQuota behavior which blocks resources before they are used
-  (e.g. for Pending Pods),
-* for LimitRanger TBD.
+* to ensure alignment with current ResourceQuota behavior that blocks resources
+  before they are used (e.g. for Pending Pods), we should do the following:
+  * for requests.[cpu|memory], Pod's max(Spec.Containers[i].Resources.Requests,
```

Comment:
Can you think of any scenarios in which only using `Spec.Containers[i].Resources.Requests` allows exceeding the quota for an extended period of time?   Using max(resources, resources allocated) only matters when resources allocated > resources, meaning a downward resize is in progress.  Since the kubelet should always admit downward resizes in requests, `resources allocated` should always quickly be lowered to `resources`.

I think just using `Spec.Containers[i].Resources.Requests` for resource quota keeps it simpler, and still provides us all of the guarantees we want from quota.

Reply:
The issue is when two or more resize (up) requests arrive quickly/concurrently at API Server when Kubelet has not had a chance to process them yet.

For example, let's consider a single-node cluster with Node capacity 8G mem, and has two running Pods P1 (req & lim: 1G, 1G), P2 (req & lim: 1G, 1G), in NS foo with ResourceQuota req & lim @ 3G, 3G respectively. 
T0: Kubelet goes offline and is restarting.
T1: P1 is resized to 2G, 2G - this is acceptable since ResourcesAllocated sum is 2G, 2G.
T2: P2 is resized to 2G, 2G - this is not acceptable but gets admitted because ResourcesAllocated is still at 2G, 2G.
T4: Kubelet completes restart, adds P1 & P2 at current (1G, 1G) level, and resizes them both to 2G, 2G, which is not OK.

If we took max(Resources.Requests, ResourcesAllocated) we would correctly reject P2 resize.

Reply:
> Can you think of any scenarios in which only using `Spec.Containers[i].Resources.Requests` allows exceeding the quota for an extended period of time? Using max(resources, resources allocated) only matters when resources allocated > resources, meaning a downward resize is in progress. Since the kubelet should always admit downward resizes in requests, `resources allocated` should always quickly be lowered to `resources`.
> 
> I think just using `Spec.Containers[i].Resources.Requests` for resource quota keeps it simpler, and still provides us all of the guarantees we want from quota.

Never mind,  I was thinking in terms of actual allocations. It is sufficient to use Spec.Containers[i].Resources.[Requests|Limits] sum across all Pods in NS for admitting resize. We are doing the right thing when actual allocations are handled in Kubelet.

Reply:
If we run into issues with this approach, we can easily switch to using max for ResourceQuota. Resolving this issue.